runtime/hrtime.cpp

```diff
@@ -0,0 +1,67 @@
+#include "runtime/datetime.h"
+
+// Moved to a separate file so we can use conditional compilation
+// with less awkwardness.
+//
+// Based on php-src/ext/standard/hrtime.c, ported to a slightly more C++ version.
+// Only posix and darwin platforms are adopted.
+
+#if defined(__APPLE__)
+#  include <mach/mach_time.h>
+#  include <string.h>
+static mach_timebase_info_data_t _timerlib_info;
+#else
+// Posix systems.
+#  include <unistd.h>
+#  include <time.h>
+#endif
+
+#include "common/wrappers/likely.h"
+
+static bool hrtime_available = false;
+constexpr time_t NANO_IN_SEC = 1'000'000'000;
+
+void init_hrtime() {
+#if defined(__APPLE__)
+  hrtime_available = (mach_timebase_info(&_timerlib_info) == KERN_SUCCESS);
+#elif defined(_POSIX_MONOTONIC_CLOCK)
+  hrtime_available = true;
+#endif
+}
+
+int64_t hrtime() {
+#if defined(__APPLE__)
```

Comment:
You can use clock_gettime for apple as well.

Reply:
Done.